SimPEG/Utils/printinfo.py

```diff
@@ -0,0 +1,281 @@
+"""
+:mod:`printinfo` -- Tools to print date, time, and version information
+======================================================================
+
+Print or return date, time, and package version information in any environment
+(Jupyter notebook, IPython console, Python console, QT console), either as
+html-table (notebook) or as plain text (anywhere).
+
+This script is an adapted version of ``empymod.printversions``, which itself
+was heavily inspired by
+
+- ``ipynbtools.py`` from https://github.com/qutip, and
+- ``watermark.py`` from https://github.com/rasbt/watermark,
+
+Always shown are the OS, number of CPU(s), ``numpy``, ``scipy``, ``SimPEG``,
+``cython``, ``properties``, ``vectormath``, ``discretize``, ``pymatsolver``,
+``sys.version``, and time/date.
+
+Additionally shown are, if they can be imported, ``IPython``, ``matplotlib``,
+and ``ipywidgets``. It also shows MKL information, if available.
+
+All modules provided in ``add_pckg`` are also shown. They have to be imported
+before ``versions`` is called.
+
+"""
+
+# Mandatory modules
+import sys
+import time
+import numpy
+import scipy
+import textwrap
+import platform
+import multiprocessing
+
+# Required packages
+import SimPEG
+import cython
+import properties
+import vectormath
+import discretize
+import pymatsolver
+
+# Optional modules
+try:
+    import IPython
+    from IPython.display import HTML, Pretty
+except ImportError:
+    IPython = False
+try:
+    import matplotlib
+except ImportError:
+    matplotlib = False
+try:
+    import ipywidgets
+except ImportError:
+    ipywidgets = False
+try:
+    import numexpr
+except ImportError:
+    numexpr = False
+try:
+    import mkl
+except ImportError:
+    mkl = False
+
+# Get mkl info from numexpr or mkl, if available
+if mkl:
+    mklinfo = mkl.get_version_string()
+elif numexpr:
+    mklinfo = numexpr.get_vml_version()
+else:
+    mklinfo = False
+
+__all__ = ['versions', 'versions_html', 'versions_text']
+
+
+def versions(mode='print', add_pckg=None, ncol=4):
```

Comment:
I wonder if changing the default here depending on if you are in IPython or not?

```python
def versions(mode='html' if IPython else 'print', add_pckg=None, ncol=4):
    ...
```

Reply:
I guess you mean `mode='HTML'`, which returns the rendered table (`mode='html'` returns the actual html-code).

You can, but I think is better not to. You would have to make sure you are in an Jupyter Notebook, because otherwise you could also be in an IPython console, and the `mode='HTML'` cannot render.

Also, what speaks for having `mode='print'` is that this looks OK in a notebook too, it would just print it as in the console. So this is the only mode that always works nicely. If there is a nice and easy trick though to ensure if you are in a Jupyter notebook then it could be a good idea, indeed.

Reply:
I had another look at this, and I could not find a good way. You could distinguish between an IPython Console and a Jupyter Notebook, but not between a Qt Console and a Jupyter Notebook. But the Qt Console cannot render the `HTML` mode either. So I think sticking with `mode='print'` as default is the safest option in whatever environment someone is using `SimPEG`.